libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -133,16 +133,17 @@ void AC_PID_2D::update_i(const Vector2f &limit)
     _pid_info_y.limit = false;
 
     Vector2f delta_integrator = (_error * _ki) * _dt;
-    if (!is_zero(limit.length_squared())) {
-        // zero delta_vel if it will increase the velocity error
+    float integrator_length = _integrator.length();
+    _integrator += delta_integrator;
+    if (!is_zero(integrator_length)) {
+        // do not let integrator increase in length if delta_integrator is in the direction of limit
         if (is_positive(delta_integrator * limit)) {
-            delta_integrator.zero();
+            _integrator.limit_length(integrator_length);
```

Comment:
```suggestion
                   if (is_positive(delta_integrator * limit) && _integrator.limit_length(integrator_length)) {
```
A change to check the return of `limit_length`

Reply:
I like that!

Reply:
Done, Thanks.